### PR TITLE
Add XGLM models

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR adds the XGLM model: [code](https://github.com/pytorch/fairseq/tree/main/examples/xglm), [paper](https://arxiv.org/abs/2112.10668)

### CODE REVIEW

File: src/transformers/models/xglm/configuration_xglm.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+XGLM_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "facebook/xglm-564M": "https://huggingface.co/facebook/xglm-564M/resolve/main/config.json",
+    # See all XGLM models at https://huggingface.co/models?filter=xglm
+}
+
+
+class XGLMConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`~XGLMModel`]. It is used to instantiate an XGLM
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/configuration_xglm.py:31]:
```suggestion
    This is the configuration class to store the configuration of a [`XGLMModel`]. It is used to instantiate an XGLM
```
Don't need the `~` anymore no cc @sgugger

File: src/transformers/models/xglm/modeling_flax_xglm.py [hunk 2]
```diff
@@ -0,0 +1,812 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Team and The HuggingFace Inc. team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_flax_xglm.py:2]: double hugging face copyright

File: src/transformers/models/xglm/configuration_xglm.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
+# coding=utf-8
+# Copyright The HuggingFace Team and The HuggingFace Inc. team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/configuration_xglm.py:2]: also one copy right is enough here (maybe facebook should be added)

File: docs/source/model_doc/xglm.mdx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+in multilingual commonsense reasoning (with +7.4% absolute accuracy improvement in 0-shot settings and +9.4% in 4-shot settings) 
+and natural language inference (+5.4% in each of 0-shot and 4-shot settings). On the FLORES-101 machine translation benchmark, 
+our model outperforms GPT-3 on 171 out of 182 translation directions with 32 training examples, while surpassing the 
+official supervised baseline in 45 directions. We present a detailed analysis of where the model succeeds and fails, 
+showing in particular that it enables cross-lingual in-context learning on some tasks, while there is still room for improvement 
+on surface form robustness and adaptation to tasks that do not have a natural cloze form. Finally, we evaluate our models 
+in social value tasks such as hate speech detection in five languages and find it has limitations similar to comparable sized GPT-3 models.*
+
+
+This model was contributed by [Suraj](https://huggingface.co/valhalla). The original code can be found [here](https://github.com/pytorch/fairseq/tree/main/examples/xglm).
```
Comment (patrickvonplaten) [inline docs/source/model_doc/xglm.mdx:39]: Maybe we can add some prompt examples here - but happy to do this in a follow-up PR

File: src/transformers/models/xglm/modeling_flax_xglm.py [hunk 5]
```diff
@@ -0,0 +1,812 @@
...
+    """
+    shifted_input_ids = jnp.roll(input_ids, 1, axis=-1)
+    shifted_input_ids = jax.ops.index_update(shifted_input_ids, (..., 0), decoder_start_token_id)
+    # replace possible -100 values in labels by `pad_token_id`
+    shifted_input_ids = jnp.where(shifted_input_ids == -100, pad_token_id, shifted_input_ids)
+
+    return shifted_input_ids
+
+
+class FlaxXGLMAttention(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_flax_xglm.py:141]: Can we use some `# Copied from ` statements here?

Comment (patrickvonplaten) [reply src/transformers/models/xglm/modeling_flax_xglm.py:141]: This looks more or less exactly like the Bart attention to me no?

File: src/transformers/models/xglm/modeling_flax_xglm.py [hunk 6]
```diff
@@ -0,0 +1,812 @@
...
+        )
+
+        attn_output = jnp.einsum("...hqk,...khd->...qhd", attn_weights, value_states)
+        attn_output = self._merge_heads(attn_output)
+        attn_output = self.out_proj(attn_output)
+
+        return attn_output, attn_weights
+
+
+class FlaxXGLMDecoderLayer(nn.Module):
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_flax_xglm.py:306]: Same here cannot use # Copied from here?

File: src/transformers/models/xglm/modeling_flax_xglm.py [hunk 7]
```diff
@@ -0,0 +1,812 @@
...
+                    all_cross_attentions += (layer_outputs[2],)
+
+        # add hidden states from the last decoder layer
+        if output_hidden_states:
+            all_hidden_states += (hidden_states,)
+
+        outputs = (hidden_states, all_hidden_states, all_self_attns, all_cross_attentions)
+        return outputs
+
+        # if not return_dict:
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_flax_xglm.py:456]: can remove this if not necessary

File: src/transformers/models/xglm/configuration_xglm.py (hunk 1 above)
Comment (patrickvonplaten) [inline src/transformers/models/xglm/configuration_xglm.py:31]:
```suggestion
    This is the configuration class to store the configuration of a [`XGLMModel`]. It is used to instantiate an XGLM
```

Think we don't need this here

Comment (patrickvonplaten) [reply src/transformers/models/xglm/configuration_xglm.py:31]: Actually not sure anymore about this (cc @sgugger) When should we use a `~` and when shouldn't we? Sometimes it seems to be included sometimes not when linking to classes

File: src/transformers/models/xglm/modeling_flax_xglm.py [hunk 8]
```diff
@@ -0,0 +1,812 @@
...
+        self.embed_tokens = nn.Embed(
+            self.config.vocab_size,
+            embed_dim,
+            embedding_init=jax.nn.initializers.normal(self.config.init_std),
+        )
+
+        # XGLM is set up so that if padding_idx is specified then offset the embedding ids by 2
+        # and adjust num_embeddings appropriately. Other models don't have this hack
+        self.offset = 2
+        # TODO: padding idx should be zero
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_flax_xglm.py:488]: Think the TODO doesn't matter here

File: src/transformers/models/xglm/modeling_flax_xglm.py [hunk 9]
```diff
@@ -0,0 +1,812 @@
...
+        return_dict: bool = True,
+        deterministic: bool = True,
+    ):
+        input_shape = input_ids.shape
+        input_ids = input_ids.reshape(-1, input_shape[-1])
+
+        inputs_embeds = self.embed_tokens(input_ids) * self.embed_scale
+
+        # embed positions
+        # position_ids = self._create_pos_ids(input_ids, position_ids)
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_flax_xglm.py:522]: delete?

File: src/transformers/models/xglm/modeling_xglm.py [hunk 10]
```diff
@@ -0,0 +1,948 @@
...
+        input_shape = inputs_embeds.size()[:-1]
+        sequence_length = input_shape[1]
+
+        position_ids = torch.arange(
+            self.padding_idx + 1, sequence_length + self.padding_idx + 1, dtype=torch.long, device=inputs_embeds.device
+        )
+        return position_ids.unsqueeze(0).expand(input_shape).contiguous()
+
+
+# Copied from transformers.models.bart.modeling_bart.BartAttention with Bart->XGLM
```
Comment (patrickvonplaten) [inline src/transformers/models/xglm/modeling_xglm.py:259]: Then we should be able to add this for Flax as well

File: tests/test_modeling_flax_xglm.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+
+    def test_use_cache_forward_with_attn_mask(self):
+        for model_class_name in self.all_model_classes:
+            config, input_ids, attention_mask = self.model_tester.prepare_config_and_inputs()
+            self.model_tester.check_use_cache_forward_with_attn_mask(
+                model_class_name, config, input_ids, attention_mask
+            )
+
+    @slow
+    def test_batch_generation(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_xglm.py:219]: nice test

File: tests/test_modeling_xglm.py [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+    @slow
+    def test_lm_generate_xglm(self):
+        self._test_lm_generate_xglm_helper()
+
+    @slow
+    def test_lm_generate_xglm_with_gradient_checkpointing(self):
+        self._test_lm_generate_xglm_helper(gradient_checkpointing=True)
+
+    @slow
+    def test_xglm_sample(self):
```
Comment (patrickvonplaten) [inline tests/test_modeling_xglm.py:418]: great tests

**Review by patrickvonplaten — APPROVED** [review]
Nice, looks very clean to me already!

File: src/transformers/models/xglm/modeling_flax_xglm.py (hunk 5 above)
Comment (patil-suraj) [reply src/transformers/models/xglm/modeling_flax_xglm.py:141]: yes, it's exactly similar to Bart attention, but bart uses `-inf` as masked value in `attention_bias`, which in XGLM causes `nan` values so we can't use `-inf` as the masked value here, so not possible to add  # Copied from here

File: src/transformers/models/xglm/modeling_flax_xglm.py (hunk 6 above)
Comment (patil-suraj) [reply src/transformers/models/xglm/modeling_flax_xglm.py:306]: yes, added it for the `__call__` method.

File: tests/test_tokenization_xglm.py [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                ".",
+            ],
+        )
+        ids = tokenizer.convert_tokens_to_ids(tokens)
+        self.assertListEqual(
+            ids,
+            [
+                value + tokenizer.fairseq_offset
+                for value in [8, 21, 84, 55, 24, 19, 7, 2, 602, 347, 347, 347, 3, 12, 66, 46, 72, 80, 6, 2, 4]
+                #                                       ^ unk: 2 + 1 = 3                  unk: 2 + 1 = 3 ^
```
Comment (sgugger) [inline tests/test_tokenization_xglm.py:110]: It points to 2 so I don't understand this comment.

**Review by sgugger — APPROVED** [review]
Thanks a lot for adding this new model!

File: tests/test_modeling_flax_xglm.py [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+                )
+                for fx_output, pt_output in zip(fx_outputs, pt_outputs_loaded):
+                    self.assert_almost_equals(fx_output[:, -1], pt_output[:, -1].numpy(), 4e-2)
+
+    @slow
+    def test_model_from_pretrained(self):
+        for model_class_name in self.all_model_classes:
+            model = model_class_name.from_pretrained("facebook/xglm-564M")
+            outputs = model(np.ones((1, 1)))
+            self.assertIsNotNone(outputs)
```
Comment (LysandreJik) [inline tests/test_modeling_flax_xglm.py:347]: Could there be an integration test here as well?

**Review by LysandreJik — APPROVED** [review]
This is great, thanks @patil-suraj!
